Alex and Happy Part 2/article.md

```diff
@@ -0,0 +1,1412 @@
+          ↙   ↘
+        1       +
+              ↙   ↘
+            2       3
+```
+
+Which was generated from this grammar:
+
+```
+exp -> num
+exp -> num "+" exp
+num -> [0-9]+
+```
+
+It means that in an expression such as `1 + 2 + 3`, a LL parser would perform substitutions by visiting the leftmost nodes first, recovering the tokens in this order:
+
+<pre>
+<i>exp</i>
+<b><i>num</i> + exp</b>
+<b>1</b> + <i>exp</i>
+1 + <b><i>num</i> + exp</b>
+1 + <b>2</b> + <i>exp</i>
+1 + 2 + <b><i>num</i></b>
+1 + 2 + <b>3</b>
+</pre>
+
+n.b.: **Bold** means the result of the derivation, and <i>italic</i> means the derivation that will be expanded next.
+
+Whilst an LR parser would do the following:
+
+<pre>
+<i>1</i> + 2 + 3
+<b>num</b> + <i>2</i> + 3
+num + <b>num</b> + <i>3</i>
+num + num + <i><b>num</b></i>
+num + <i>num + <b>exp</b></i>
+<i>num + <b>exp</b></i>
+<b>exp</b>
+</pre>
+
+For more information on leftmost and rightmost derivations, see [LL and LR Parsing Demystified](https://blog.reverberate.org/2013/07/ll-and-lr-parsing-demystified.html).
+
+Another difference is that Alex is a lexer generator and Happy a parser generator.
+It means that instead of writing Haskell files (`.hs`) like you do with Megaparsec, you will write Alex (`.x`) and Happy (`.y`) files, which in turn will automatically generate Haskell files with implementations of the lexer and parser recognizing the described grammar.
+
+Below are some pros and cons of Happy in comparison to Megaparsec:
+
+Pros:
+* Ambiguous grammars are reported to the user with parser generators, allowing a greater trust in the parser's output and saving the debugging headache.
+* Bottom-up parsers can handle left-recursion, while top-down can't.
+* The performance will be more predictable, as Megaparsec can cause exponential time complexity with the use of `try`. In addition, Happy's "static" generated LALR algorithm has better known complexity constraints.
+
+Cons:
+* Less idiomatic, as you need to write Alex and Happy grammars instead of Haskell code.
+* Less flexible since you can't touch the algorithm on how Happy works. Meanwhile, you can easily create your functions for parser combinators.
+* Dealing with operator precedence and associativity requires more labor. There are standard tools to deal with them, but once you need something that is non-standard, you'll need to write more non-trivial parser rules by hand.
+
+Comparing the implementations of both parsing techniques is beyond the scope of this post. Still, if you are interested, you may read the parser combinators article linked above.
+
+## How it works
+
+An in-depth explanation of how an LR(1) parser works is well beyond the scope of this tutorial. For this, we recommend you read our article on [how to Implement an LR(1) Parser](https://serokell.io/blog/how-to-implement-lr1-parser).
+We will, however, introduce a few basic concepts required to understand this article better.
+
+### Terminals, non-terminals, and production rules
```

Comment:
![image](https://user-images.githubusercontent.com/35335196/173025337-846e28d5-4c08-417d-a3c8-d5d58067c3f4.png)
https://nadia-polikarpova.github.io/cse130-web/lectures/06-parsing.html

Reply:
Added some simplified sentences according to it. Hope it makes it a bit easier to understand.